### ISSUE TITLE
Allow downloading remote code of subcomponents

### ISSUE THREAD

**Issue by patrickvonplaten** [huggingface/diffusers#4966]
It could make a lot of sense to allow people to download remote code along side the model checkpoint, e.g. as proposed here: https://huggingface.co/tomerkeren42/DiffusionExperimental-5/blob/main/unet_model/modeling_unet.py 

We could add a new mechanism to the `model_index.json`, possibly coupled with `trust_remote_code=True` for this

**Comment by sayakpaul** [comment]
@patrickvonplaten the link leads to 404. Could you maybe elaborate a bit here to explain the use-cases and the kind of APIs you'd like to see?

**Comment by patrickvonplaten** [comment]
Ah yeah that was made private - in short we might want to allow downloading code that resides inside model repos just like it's done for `transformers`: https://huggingface.co/docs/transformers/v4.33.0/en/custom_models

**Comment by sayakpaul** [comment]
Nice! Makes a lot of sense! 

What do we need for this?

* `trust_remote_code` in `from_pretrained()` for [`ModelMixin`](https://github.com/huggingface/diffusers/blob/5fd42e5d6101eb871837399db6f5015cf2c319b5/src/diffusers/models/modeling_utils.py#L177) we have in `diffusers`
* Registration for the auto classes (e.g., `SayakUNet2DConditionModel.register_for_auto_class()`

I don't think we should extend this to pipelines as there's already a flexible way to allow the execution of custom pipelines. 

Am I missing out on anything here?

**Comment by patrickvonplaten** [comment]
Yes `trust_remote_code` as a flag to `from_pretrained(...)` and then I'm not yet sure about the best way to register the model since we don't have auto classes

**Comment by sayakpaul** [comment]
@patrickvonplaten I have a very minimal solution. Let me know if that works:

First, define your model:

```python
from diffusers.models.modeling_utils import ModelMixin
from diffusers.configuration_utils import ConfigMixin, register_to_config 
import torch.nn as nn
import torch 
from typing import Union, Optional, Dict, Any, Tuple


class SayakUNet2DConditionModel(ModelMixin, ConfigMixin):
    @register_to_config
    def __init__(self, in_channels: int, out_channels: int):
        super().__init__()
        
        self.in_channels = in_channels
        self.out_channels = out_channels
        self.conv_in = nn.Conv2d(
            in_channels, out_channels, kernel_size=3, padding=1
        )

    def forward(self, 
        sample: torch.FloatTensor,
        timestep: Union[torch.Tensor, float, int],
        encoder_hidden_states: torch.Tensor,
        class_labels: Optional[torch.Tensor] = None,
        timestep_cond: Optional[torch.Tensor] = None,
        attention_mask: Optional[torch.Tensor] = None,
        cross_attention_kwargs: Optional[Dict[str, Any]] = None,
        added_cond_kwargs: Optional[Dict[str, torch.Tensor]] = None,
        down_block_additional_residuals: Optional[Tuple[torch.Tensor]] = None,
        mid_block_additional_residual: Optional[torch.Tensor] = None,
        encoder_attention_mask: Optional[torch.Tensor] = None,
        return_dict: bool = True
    ):
        return (self.conv_in(sample), )
```

Push it to Hub:

```python
sayak_unet = SayakUNet2DConditionModel(4, 4)
sayak_unet.push_to_hub("my_unet")
```

Create a `RemoteModelLoader`:

```python
import torch
import importlib
import os
from huggingface_hub import hf_hub_download


DEFAULT_MODULE_FILENAME = "modeling_custom_unet.py"
MODEL_WEIGHTS_NAME = "diffusion_pytorch_model.safetensors"

class RemoteModelLoader:
    @classmethod
    def from_pretrained(cls, repo_id, model_class_name, module_filepath=None, weights_name=None):
        # Download the model code from S3
        module_path = hf_hub_download(repo_id=repo_id, filename=module_filepath)
        
        # Import the model code as a Python module
        module_name = module_path.split("/")[-1].replace(".py", "")  # Choose a name for the imported module
        spec = importlib.util.spec_from_file_location(module_name, module_path)
        model_module = importlib.util.module_from_spec(spec)
        spec.loader.exec_module(model_module)
        
        # Initialize the model class
        model = getattr(model_module, model_class_name) 
        
        # Load the model weights
        model.from_pretrained(repo_id)
        
        return model
```

Use:

```python
my_unet = RemoteModelLoader.from_pretrained(
    "sayakpaul/my_unet", model_class_name="SayakUNet2DConditionModel", module_filepath=DEFAULT_MODULE_FILENAME
)
```

**Comment by patrickvonplaten** [comment]
Hmm, I need to think about this a bit more. It's probably not that easy to implement. We ideally don't want to force the user to pass `"SayakUNet2DConditionModel"` and instead loading should work just by doing:

```py
pipe = DiffusionPipeline.from_pretrained("sayakpaul/my_pipeline", trust_remote_code=True)  # my_pipeline has a subfolder called `"my_unet"`
```

or/and:

```py
unet = AutoModel.from_pretrained("sayakpaul/my_unet", trust_remote_code=True) 
```

We probably would have to first create a `AutoModel` class that allows to download diffusers models and then one could register new classes on this `AutoModel` class.

What do you think about the idea of a `AutoModel` class?

**Comment by sayakpaul** [comment]
I did think about `AutoModel` and thoughts its implications in the context of the library. 

Unlike `transformers`, I don't think having `AutoModel` would have a good amount of impact in `diffusers`. In `diffusers`, the nature of being explicit is more pronounced I think. 

`transformers` exposes many "model" types and having auto classes for them makes sense. Whereas, in `diffusers`, we have many "pipeline" types (a pipeline is a composition of different models) and for that, we already have auto classes. So, with that being the main reason, I think having things a bit more explicitly done (like I showed above) would be better from the perspective of maintenance and user-friendliness. 

I think `diffusers` users are okay being more explicit. And if you think about it in retrospect, the solution I proposed doesn't deviate too much (in terms of the developer experience) compared to how it's done in `transformers`.
